File: topic/binary_tree_level_order_traversal/binary_tree_level_order_traversal.go
package binary_tree_level_order_traversal

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func LevelOrder(root *TreeNode) [][]int {
	if root == nil {
		return nil
	}
	result := [][]int{}
	queue := []TreeNode{*root}
	tmpQueue := []TreeNode{}
	tmp := []int{}
	for len(queue) != 0 || len(tmpQueue) != 0 {
		if len(queue) == 0 {
			queue = tmpQueue
			result = append(result, tmp)
			tmpQueue = nil
			tmp = nil
		} else {
			tmp = append(tmp, queue[0].Val)
			if queue[0].Left != nil {
				tmpQueue = append(tmpQueue, *queue[0].Left)
			}
			if queue[0].Right != nil {
				tmpQueue = append(tmpQueue, *queue[0].Right)
			}
			queue = queue[1:]
		}
	}

	if tmp != nil {
		result = append(result, tmp)
	}

	return result
}

File: topic/binary_tree_level_order_traversal/cpp/binary_tree_level_order_traversal.cpp
#include <vector>
#include <queue>
using namespace std;

struct TreeNode
{
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode() : val(0), left(nullptr), right(nullptr) {}
	TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
	TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
	vector<vector<int> > levelOrder(TreeNode *root)
	{
		vector<vector<int> > result;
		queue<TreeNode *> queue;

		if (root == nullptr)
			return result;

		queue.push(root);

		while (!queue.empty())
		{
			result.push_back(vector<int>{});
			int n = queue.size();

			for (int i = 0; i < n; i++)
			{
				result.back().push_back(queue.front()->val);

				if (queue.front()->left != nullptr)
					queue.push(queue.front()->left);

				if (queue.front()->right != nullptr)
					queue.push(queue.front()->right);

				queue.pop();
			}
		}

		return result;
	}
};